File: Data Structures & Algorithms/subtree-of-a-binary-tree/submission-6.js
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     constructor(val = 0, left = null, right = null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    /**
     * @param {TreeNode} root
     * @param {TreeNode} subRoot
     * @return {boolean}
     */
    isSubtree(root, subRoot) {
        if (!subRoot) return true;
        const findNode = (node) => {
            if (!node) return false;
            if (node.val === subRoot.val) {
                if (this.dfs(node, subRoot)) return true;
            }
            return findNode(node.left) || findNode(node.right);
        }
        return findNode(root);
    }
    
    dfs(n1, n2) {
        if (!n1 && !n2) return true;
        if (!n1 || !n2 || n1.val !== n2.val) return false;
        return this.dfs(n1.left, n2.left) && this.dfs(n1.right, n2.right);
    }
}

File: Data Structures & Algorithms/valid-binary-search-tree/submission-6.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        if not root:
            return True
        # [res, [tmpMin, tmpMax]]
        defaultCandi = [float('inf'), -float('inf')]
        def dfs(n):
            if not n:
                return [True, defaultCandi]

            [isValidL, [minL, maxL]] = dfs(n.left)
            [isValidR, [minR, maxR]] = dfs(n.right)
            if not isValidL or not isValidR or not (maxL < n.val < minR):
                return [False, defaultCandi]

            return [True, [min(minL, n.val), max(maxR, n.val)]]
            
        return dfs(root)[0]